File: 02/02.jl
timestamp_nano() = Int(round(time() * 1e9))

part1(data)::Int = sum(
    (you - 'W') + # ASCII hax
    ifelse(opp + ('X' - 'A') == you,
        3,
        ifelse((opp == 'A' && you == 'Y') || (opp == 'B' && you == 'Z') || (opp == 'C' && you == 'X'),
            6,
            0 # empty case
        )
    ) for (opp, _, you) in data
)

part2(data)::Int = sum(
    ifelse(you == 'X', # loss
        ifelse(opp == 'A',
            3,
            ifelse(opp == 'B', 1, 2)
        ),
        ifelse(you == 'Y', # draw
            3 + (opp - 'A' + 1),
            6 + ifelse(opp == 'A', # win
                2,
                ifelse(opp == 'B',
                    3,
                    1
                )
            )
        )
    ) for (opp, _, you) in data
)

get_data() = open("02/input.txt") do in_file
    split(read(in_file, String), "\n")
end

start = timestamp_nano()

data = get_data()

println("part1: ", part1(data))
println("part2: ", part2(data))

end_ = timestamp_nano()
println("time: ", round((end_ - start) / 1e6, digits=3), "ms")

File: 02/02.py
from timing_util import Timing


def get_data(content: str) -> list[list[str]]:
    return [game.split(' ') for game in content.splitlines()]


def part1(data: list[list[str]]) -> int:
    diff = ord('X') - ord('A')
    wins = {('A', 'Y'), ('B', 'Z'), ('C', 'X')}
    offset = ord('W')

    score = 0
    for opp, you in data:
        if (opp, you) in wins:
            score += 6
        elif ord(opp) + diff == ord(you):
            score += 3

        score += ord(you) - offset
    return score


def part2(data: list[list[str]]) -> int:
    score = 0
    for opp, you in data:
        if you == 'X':  # win
            if opp == 'A':
                score += 3
            elif opp == 'B':
                score += 1
            else:
                score += 2
        elif you == 'Y':  # draw
            score += 3 + (ord(opp) - ord('A') + 1)
        else:  # loss
            score += 6
            if opp == 'A':
                score += 2
            elif opp == 'B':
                score += 3
            else:
                score += 1
    return score


if __name__ == '__main__':
    with Timing():
        with open('02/input.txt') as in_file:
            data = get_data(in_file.read())

        print(f'part1: {part1(data)}')
        print(f'part2: {part2(data)}')
